### ISSUE TITLE
Mosaic crystal in Laue geometry

### ISSUE THREAD

**Issue by camsbu** [kklmn/xrt#136]
Hello,

I'm trying to simulate a mosaic C<111> crystal in Laue geometry using reflection on the (11-1) plane. I'm probably not doing this the best way - so far, I have been using the LauePlate class, and adjusted the pitch and asymmetry angles to get to the (hopefully) right setup: with the angle between the (111) and (11-1) planes being arccos(1/3), I'm using a LauePlate with pitch = Bragg angle + arccos(1/3) and alpha = arccos(1/3) - pi/2. 
My setup also includes a screen centered on and perpendicular to the diffracted beam, and I also added an aperture in front of the screen to block anything but the diffracted beam, as in test_mosaic_crystal_thin.py.

This seems to be working well with a perfect crystal (no mosaicity) and I'm getting a reasonable rocking curve (right aspect, peak reflectivity, width and all of the rays are going through the aperture). When switching to a mosaic crystal however, no rays are coming out of the aperture. When I tried removing the aperture, a very (too?) intense beam appears, at the position where I would have expected the diffracted beam to be, meaning it should not have been blocked by the aperture.

Would you have any insight on the above? I'm probably missing something obvious or not doing things correctly. Happy to provide more details or plots if that helps.

Many thanks in advance for the help!

**Comment by kklmn** [comment]
Hi,
Your exercise with the aperture is unclear. Please post your script and refer your explanations to it.

**Comment by camsbu** [comment]
Thanks very much for the quick reply, so sorry for the confusion. I realized I was not plotting the right beam at some point which was causing some of the errors, but there are still some things that are puzzling to me. Here is an example of what I have been using:

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os, sys; sys.path.append(os.path.join('..', '..', '..'))  # analysis:ignore
import numpy as np
import math
import matplotlib.pyplot as plt
import xrt.backends.raycing as raycing
import xrt.backends.raycing.sources as rs
import xrt.backends.raycing.apertures as ra
import xrt.backends.raycing.oes as roe
import xrt.backends.raycing.run as rr
import xrt.backends.raycing.materials as rm
import xrt.plotter as xrtp
import xrt.runner as xrtr
import xrt.backends.raycing.screens as rsc


# ----------------------------------------------------------------------------
# Define source parameters
# ----------------------------------------------------------------------------

# Spectrum
E0 = 8048. # eV
deltaE = 20. # eV, FWHM

# Number of rays in the simulation (= NRays * NRepeats)
NRays = 100000
NRepeats = 1

# Beam standard deviation (mm)
Beam_stdv = 0.025/2.355  # mm  25 microns FWHM beam size

# Beam divergence (rad, sigma)
Beam_div_rad = 0.  # rad


# ----------------------------------------------------------------------------
# Create crystals:
# ----------------------------------------------------------------------------

a_C = 3.56683
rho_C = 3.51

# C(111)
geom_C111 = 'Laue reflected'
C111_hkl = (1,-1,1)
d_C111 = a_C/math.sqrt(sum(HKL**2 for HKL in C111_hkl))
mosaicity_C111_deg = 0.0055/2.355
mosaicity_C111_rad = mosaicity_C111_deg*np.pi/180.
t = 0.1
angle_planes_C111_rad = np.arccos(1./3.)

xtal_C111 = rm.CrystalDiamond(hkl=C111_hkl, elements='C', t=t, rho=rho_C, d=d_C111, 
                            geom=geom_C111, mosaicity=mosaicity_C111_rad)

xtal_C111_perfect = rm.CrystalDiamond(hkl=C111_hkl, elements='C', t=t, rho=rho_C, 
                                      d=d_C111, geom=geom_C111)


Bragg_C111_rad = xtal_C111_perfect.get_Bragg_angle(E0) - xtal_C111_perfect.get_dtheta(E0)

alpha_C111_rad = angle_planes_C111_rad - np.pi/2.
pitch_C111_rad = Bragg_C111_rad + angle_planes_C111_rad


# ----------------------------------------------------------------------------
# Setup geometry
# ----------------------------------------------------------------------------

Dist_Source_Crystal = 100.
h = 152.5
L = h/np.tan(2.*Bragg_C111_rad)


# ----------------------------------------------------------------------------
# Create "beamline"
# ----------------------------------------------------------------------------
def build_beamline(nrays=NRays):
    # Create beamline
    beamLine = raycing.BeamLine(height=0)
    
    # ------------------------
    # Source:
    # ------------------------
    # Create source (geometrical gaussian source)
    beamLine.Source = rs.GeometricSource(bl=beamLine, name='GeometricSource', 
        center=(0,-Dist_Source_Crystal,0), 
        nrays=nrays,
        distx='normal', dx=Beam_stdv, disty='None', dy=0., 
        distz='normal', dz=Beam_stdv,
        distxprime='normal', dxprime=Beam_div_rad, 
        distzprime='normal', dzprime=Beam_div_rad, 
        distE='normal', energies=(E0,deltaE/2.355),
        polarization='vertical')
    
    # ------------------------
    # Crystal
    # ------------------------
    beamLine.C111XTal = roe.LauePlate(bl=beamLine, name='C111', 
                                      center=[0,0,0], pitch=pitch_C111_rad, 
                                      material=xtal_C111, alpha=alpha_C111_rad)
    
    # ------------------------
    # Screen (perpendicular to the diffracted beam)
    # ------------------------
    # Aperture to select the reflected beam only
    beamLine.aperture = ra.RectangularAperture(beamLine, opening=[-50, 50, -50, 50], 
                                               center=[0, L, h], 
                                               z=[0, -np.sin(2.*Bragg_C111_rad), np.cos(2.*Bragg_C111_rad)])
    
    # Screen perpendicular to the diffracted beam
    beamLine.screen = rsc.Screen(beamLine, 'screen', (0, L, h), 
                                  z=[0, -np.sin(2.*Bragg_C111_rad), np.cos(2.*Bragg_C111_rad)])
    # Screen perpendicular to the direct beam
    beamLine.screenPerp = rsc.Screen(beamLine, 'screenPerp', (0, L, 0))
    
    return beamLine


def run_process(beamLine):
    # ---- SOURCE ----
    # Create source beam
    beamSource = beamLine.Source.shine()
    
    # Reflect beam on C111 crystal
    beamC111Global, beamC111Local = beamLine.C111XTal.reflect(beamSource)
    
    # Propagate reflected beam through aperture
    beamC111Reflected = beamLine.aperture.propagate(beamC111Global)
    # beamC111Reflected = beamC111Global
    beamScreen = beamLine.screen.expose(beamC111Reflected)
    beamScreenPerp = beamLine.screenPerp.expose(beamC111Reflected)
    
    outDict = {"beamSource": beamSource, "beamC111Global": beamC111Global,
               "beamC111Local": beamC111Local, "beamC111Reflected":beamC111Reflected,
               "beamScreen": beamScreen, "beamScreenPerp": beamScreenPerp}
    return outDict


rr.run_process = run_process



# ----------------------------------------------------------------------------
# Define plots
# ----------------------------------------------------------------------------
def define_plots(beamLine):
    plots = []
    
    plot00 = xrtp.XYCPlot('beamSource', (1,),
        xaxis=xrtp.XYCAxis(r'$x$', 'mm', limits=[-0.05, 0.05], fwhmFormatStr = '%.4f'),
        yaxis=xrtp.XYCAxis(r'$z$', 'mm', limits=[-0.05, 0.05], fwhmFormatStr = '%.4f'),
        ePos=1, title='beamSource')
    plot00.caxis.limits = [8018., 8078.]
    plots.append(plot00)
    
    plot01 = xrtp.XYCPlot('beamScreen', (1,),
        xaxis=xrtp.XYCAxis(r'$x$', 'mm', limits=[-0.05, 0.05], fwhmFormatStr = '%.4f'),
        yaxis=xrtp.XYCAxis(r'$z$', 'mm', limits=[-0.05, 0.05], fwhmFormatStr = '%.4f'),
        ePos=1, title='beamScreen')
    plot01.caxis.limits = [8018., 8078.]
    plots.append(plot01)
    
    plot02 = xrtp.XYCPlot('beamScreenPerp', (1,),
        xaxis=xrtp.XYCAxis(r'$x$', 'mm', fwhmFormatStr = '%.4f'),
        yaxis=xrtp.XYCAxis(r'$z$', 'mm', fwhmFormatStr = '%.4f'),
        ePos=1, title='beamScreenPerp')
    plot02.caxis.limits = [8018., 8078.]
    plots.append(plot02)
    
    return plots



# ----------------------------------------------------------------------------
# Define scan
# ----------------------------------------------------------------------------
def plot_generator_pitch_Laue(plots, beamLine):
    dtheta_Laue_array = np.linspace(-0.01, 0.01, num=51)
    Intensity = np.zeros(len(dtheta_Laue_array))
    for idtheta, dtheta in enumerate(dtheta_Laue_array):
        print("Step {} of {}".format(idtheta+1, len(dtheta_Laue_array)))
        beamLine.C111XTal.pitch = pitch_C111_rad + dtheta*np.pi/180.
        for plot in plots:
            try:
                plot.textPanel.set_text(
                    u'\n$dθ_{Laue} = $'+'{:.1f}'.format(dtheta))
            except:
                pass
        yield
        Intensity[idtheta] = plots[1].intensity
        print("Intensity after Laue BS: ", Intensity[idtheta])
    plt.figure('Laue pitch scan')
    plt.plot(dtheta_Laue_array, Intensity)
    plt.xlabel(u'θ-θ$_{Bragg}$ (deg)')
    plt.ylabel('Intensity on the detector after Laue crystal')

def main():
    beamLine = build_beamline()
    plots = define_plots(beamLine)
    xrtr.run_ray_tracing(plots, repeats=NRepeats, beamLine=beamLine, 
                            generator=plot_generator_pitch_Laue
                         )

#this is necessary to use multiprocessing in Windows, otherwise the new Python
#contexts cannot be initialized:
if __name__ == '__main__':
    main()
```

I tried several variations of the above script, including:

1. perfect crystal (`mosaicity_C111_deg = 0.`), monochromatic beam (`deltaE = 0.`), removing the aperture in front of the screens (uncommented `beamC111Reflected = beamC111Global` and commented `beamC111Reflected = beamLine.aperture.propagate(beamC111Global)` in `run_process`). The rocking curve looks reasonable and the beam is  at its expected position at the center of `screen` (the screen perpendicular to the diffracted beam at z=152.5mm). 
![image](https://user-images.githubusercontent.com/125786558/220042433-e34993e8-631e-4278-a3ae-1efe89562746.png)
![image](https://user-images.githubusercontent.com/125786558/220042456-46bf5c49-f261-4b32-97e2-1241feb4b8a9.png)

2. same as 1. (monochromatic beam, no aperture) with mosaic crystal (`mosaicity_C111_deg = 0.0055/2.335`). The rocking curve does not look right (constant reflectivity vs. angle, too high reflectivity value). There is almost nothing where the diffracted beam is expected to be (center of `screen`) and a very intense beam on axis (center of `screenPerp`).
![image](https://user-images.githubusercontent.com/125786558/220042641-c3cc7511-49db-4fc3-b574-76eee22caf02.png)
![image](https://user-images.githubusercontent.com/125786558/220042661-a9376c18-869a-40b4-b729-ca532441882e.png)

3. same as 2. (monochromatic beam, mosaic crystal) with aperture in front of the screens (commented `beamC111Reflected = beamC111Global` and uncommented `beamC111Reflected = beamLine.aperture.propagate(beamC111Global)` in `run_process`): close to 0% reflectivity, almost no ray go through the aperture
![image](https://user-images.githubusercontent.com/125786558/220041874-d2e03387-3a3b-4759-ad14-4f14cbf57d71.png)
![image](https://user-images.githubusercontent.com/125786558/220041904-6817f1a0-3d9d-4b32-985b-2432cb30070a.png)

4. same as 2. (mosaic crystal, no aperture) with non-monochromatic beam (`deltaE = 20.`). Again, the beam is on axis, not where it was expected to be, the reflectivity is very high and constant vs. crystal pitch. The spectrum is identical to the source's.  
![image](https://user-images.githubusercontent.com/125786558/220042789-5056f424-2f9e-461e-8285-7407f511b47b.png)
![image](https://user-images.githubusercontent.com/125786558/220042812-9efa0e68-abd6-4750-bef0-ad03afcbaca6.png)

To summarize, it seems to me that I can get good results with a perfect crystal + screen alone (no aperture).
But how about the mosaic crystal? Given the beam intensity and position, I was under the impression that the simulations were showing a superposition of the diffracted and transmitted beams, which is why I started using an aperture, but that does not seem to be the case. Is there anything I am doing incorrectly?

Many thanks for your time and help! Let me know if any of the above is not clear or if you need more information.

**Comment by kklmn** [comment]
Hi,

I am not sure the Laue case of mosaic crystals has ever been tested. I don't have any paper to reference my calculations against. If you know a pepper or have experimental results, I can work on the case in some not-so-distant future.

About your attempts, if you see a black 2D plot when one of the two 1D histograms is non-zero, like in your first plot under (2.), remove the plot limits of the other 1D axis (that shows no histogram). The beam must be elsewhere, not where you expect it. Is its position correct is another question.
